lib/charms/loki_k8s/v1/loki_push_api.py

```diff
@@ -2313,6 +2387,120 @@ def _containers(self) -> Dict[str, Container]:
         return {cont: self._charm.unit.get_container(cont) for cont in self._logs_scheme.keys()}
 
 
+class LogForwarder(Object):
+    """Forward the standard outputs of all workloads operated by a charm to one or multiple Loki endpoints."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        *,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        loki_endpoint_getter: Optional[Callable[[str], List[str]]] = None,
+    ):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self.topology = JujuTopology.from_charm(charm)
+        self.endpoints_getter: Callable = (
+            self._fetch_endpoints if loki_endpoint_getter is None else loki_endpoint_getter
+        )
+        self.loki_endpoints = self.endpoints_getter(relation_name)
+
+        on_relation = self._charm.on[self._relation_name]
+        self.framework.observe(on_relation.relation_joined, self._on_logging_relation_joined)
+        self.framework.observe(on_relation.relation_changed, self._on_logging_relation_changed)
+        self.framework.observe(on_relation.relation_departed, self._on_logging_relation_departed)
+        self.framework.observe(on_relation.relation_broken, self._on_logging_relation_broken)
+
+    def _on_logging_relation_joined(self, _):
+        self.enable()
+
+    def _on_logging_relation_changed(self, _):
+        self.enable()
+
+    def _on_logging_relation_departed(self, _):
+        self.disable()
+
+    def _on_logging_relation_broken(self, _):
+        self.disable()
+
+    def enable(self):
```

Comment:
Perhaps enable/disable should also be private? There could be a use case for manually managing the status of this patch but until it's clear I say let's make this private. It's easier to make it public again if we change our mind than do the opposite.